include/nbl/video/utilities/IUtilities.h

```diff
@@ -336,22 +371,39 @@ class NBL_API2 IUtilities : public core::IReferenceCounted
                 // get allocation size
-                const uint32_t allocationSize = getAllocationSizeForStreamingBuffer(size,m_allocationAlignment,maxFreeBlock,optimalTransferAtom);
+                uint32_t allocationSize = getAllocationSizeForStreamingBuffer(size,m_allocationAlignment,maxFreeBlock,optimalTransferAtom);
                 // make sure we dont overrun the destination buffer due to padding
-                const uint32_t subSize = core::min(allocationSize,size);
+                uint32_t subSize = core::min(allocationSize,size);
                 // cannot use `multi_place` because of the extra padding size we could have added
                 uint32_t localOffset = StreamingTransientDataBufferMT<>::invalid_value;
                 m_defaultUploadBuffer.get()->multi_allocate(std::chrono::steady_clock::now()+std::chrono::microseconds(500u),1u,&localOffset,&allocationSize,&m_allocationAlignment);
                 // copy only the unpadded part
                 if (localOffset!=StreamingTransientDataBufferMT<>::invalid_value)
                 {
-                    const void* dataPtr = reinterpret_cast<const uint8_t*>(data) + uploadedSize;
-                    memcpy(reinterpret_cast<uint8_t*>(m_defaultUploadBuffer->getBufferPointer()) + localOffset, dataPtr, subSize);
+                    const uint32_t bytesWritten = callback(reinterpret_cast<uint8_t*>(m_defaultUploadBuffer->getBufferPointer())+localOffset,uploadedSize,subSize);
+                    assert(bytesWritten>0 && bytesWritten<=subSize);
+                    // Highly Experimental, enable at own risk!
+                    if constexpr (false)
+                    // Reclaim the unused space if both the used part and the unused part are large enough to be their own independent free blocks in the allocator
+                    if (const uint32_t unusedSize=subSize-bytesWritten; bytesWritten>=minBlockSize && unusedSize>=minBlockSize)
+                    {
+                        const uint32_t unusedOffset = localOffset+bytesWritten;
+                        m_defaultUploadBuffer.get()->multi_deallocate(1u,&unusedOffset,&unusedSize);
+                        allocationSize = bytesWritten;
+                    }
+                    subSize = bytesWritten;
                 }
                 else
                 {
+                    if (!flushRanges.empty())
+                    {
+                        m_device->flushMappedMemoryRanges(flushRanges);
+                        flushRanges.clear();
+                    }
                     const auto completed = nextSubmit.getFutureScratchSemaphore();
                     nextSubmit.overflowSubmit(scratch);
+                    // first submit we respect whatever stages the user had (maybe they wanted to be notified of the completion of `nextSubmit.prevCommandBuffers`
+                    nextSubmit.scratchSemaphore.stageMask = asset::PIPELINE_STAGE_FLAGS::COPY_BIT;
```

Comment:
@Erfan-Ahmadi this is the sync bug I fixed with regard to timeline semaphore signalling

Reply:
I also fixed it on buffer downloads and image updates